targets/TARGET_STM/lp_ticker.c

```diff
@@ -217,42 +258,102 @@ uint32_t lp_ticker_read(void)
 
-    lp_ticker_clear_interrupt();
+    /* CMPOK is set by hardware to inform application that the APB bus write operation to the
+     * LPTIM_CMP register has been successfully completed.
+     * Any successive write before the CMPOK flag be set, will lead to unpredictable results
+     * We need to prevent to set a new comparator value before CMPOK flag is set by HW */
+    if (lp_cmpok == false) {
+        /* if this is not safe to write, then delay the programing to the
+         * time when CMPOK interrupt will trigger */
+        lp_delayed_prog = true;
+    } else {
+        timestamp_t last_read_counter = lp_ticker_read();
+        lp_ticker_clear_interrupt();
+
+        /*  HW is not able to trig a very short term interrupt, that is
+         *  not less than few ticks away (LP_TIMER_SAFE_GUARD). So let's make sure it'
+         *  s at least current tick + LP_TIMER_SAFE_GUARD */
+        for(uint8_t i = 0; i < LP_TIMER_SAFE_GUARD; i++) {
+            if (LP_TIMER_WRAP(last_read_counter  + i) == timestamp) {
+                timestamp = LP_TIMER_WRAP(timestamp + LP_TIMER_SAFE_GUARD);
+            }
+        }
+        /* Then check if this target timestamp is not in the past, or close to wrap-around */
+        if((timestamp < last_read_counter) && (last_read_counter <= (0xFFFF - LP_TIMER_SAFE_GUARD))) {
+            /*  Workaround, because limitation */
+            __HAL_LPTIM_COMPARE_SET(&LptimHandle, ~0);
+        } else {
+            /*  It is safe to write */
+            __HAL_LPTIM_COMPARE_SET(&LptimHandle, timestamp);
+        }
 
-    NVIC_EnableIRQ(LPTIM1_IRQn);
+        /* We just programed the CMP so we'll need to wait for cmpok before
+         * next programing */
+        lp_cmpok = false;
+        /*  Prevent from sleeping after compare register was set as we need CMPOK
+         *  interrupt to fire (in ~3x30us cycles) before we can safely enter deep sleep mode */
+        if(!sleep_manager_locked) {
+            sleep_manager_lock_deep_sleep();
+            sleep_manager_locked = true;
+        }
+    }
+    core_util_critical_section_exit();
 }
 
 void lp_ticker_fire_interrupt(void)
 {
+    core_util_critical_section_enter();
     lp_Fired = 1;
+    /* In case we fire interrupt now, then cancel pending programing */
+    lp_delayed_prog = false;
     irq_handler = (void (*)(void))lp_ticker_irq_handler;
     NVIC_SetPendingIRQ(LPTIM1_IRQn);
     NVIC_EnableIRQ(LPTIM1_IRQn);
+    core_util_critical_section_exit();
 }
 
 void lp_ticker_disable_interrupt(void)
 {
+    core_util_critical_section_enter();
+
+    if(!lp_cmpok) {
+        while (__HAL_LPTIM_GET_FLAG(&LptimHandle, LPTIM_FLAG_CMPOK) == RESET) {
+        }
+        __HAL_LPTIM_CLEAR_FLAG(&LptimHandle, LPTIM_FLAG_CMPOK);
+        lp_cmpok = true;
+    }
+    /*  now that CMPOK is set, allow deep sleep again */
+    if(sleep_manager_locked) {
+        sleep_manager_unlock_deep_sleep();
+        sleep_manager_locked = false;
+    }
+    irq_handler = NULL;
```

Comment:
irq_handler is used to make sure that no user callback will be called after the calls to lp_ticker_disable_interrupt(). I agree this may not be needed, but I would maybe propose to do this optimization (and maybe others) in a future PR rather than this one.

Reply:
As other depending Prs are still under review, I added the few optimization changes as well